replica/table.cc

```diff
@@ -147,6 +147,25 @@ table::find_row(schema_ptr s, reader_permit permit, const dht::decorated_key& pa
     });
 }
 
+void
+table::add_memtables_to_reader_list(std::vector<flat_mutation_reader_v2>& readers,
+        const schema_ptr& s,
+        const reader_permit& permit,
+        const dht::partition_range& range,
+        const query::partition_slice& slice,
+        const io_priority_class& pc,
+        const tracing::trace_state_ptr& trace_state,
+        streamed_mutation::forwarding fwd,
+        mutation_reader::forwarding fwd_mr) const {
+    readers.reserve(readers.size() + _compaction_group->memtables()->size() + 1);
+
+    for (auto&& mt : *_compaction_group->memtables()) {
+        if (auto reader_opt = mt->make_flat_reader_opt(s, permit, range, slice, pc, trace_state, fwd, fwd_mr)) {
+            readers.emplace_back(std::move(*reader_opt));
+        }
```

Comment:
For point readers, we should only use one compaction group. For scans, we should open the readers one by one to avoid excessive merging.

Reply:
I'm taking it partially back. Single-point reads are really easy and more important, so let's do them. We can wait with range reads.

However, please verify (via tracing) that a full scan still opens sstables incrementally. For example, reading the very first page should only create readers for sstables in the first compaction group.

Reply:
I think with compound_sstable_set, it won't work. compound_sstable_set doesn't assume there is any ordering between the sstable sets, and always selects from all its inputs. So a scan from 32 compaction groups will merge from 32 at least readers.

If we had a tablet_sstable_set, it would select from tablets sequentially.

Reply:
tablet_sstable_set may be really necessary for dynamic tablets, as once tablets are split, disjointness is broken at sstable level. so the set makes sure that only one compaction group will feed data only for the range it owns, avoiding read amplification too.